ReadMe
milter-blockdomain

+------------------+
|   INTRODUCTION   |
+------------------+

The milter-blockdomain is a simple milter written in C to maintain
thousands of block domains in simple plain expression.

  - Separate configuration file and white/grey/black domain list file.
  - Domain list file is auto-loaded when it is updated.
    No restarting is required.
  - Check FQDN of dns reverse lookup name, helo, envelope from and/or
    envelope recipient.
    One domain entry in the domain list is applied to them.
  - Domain entry is a simple plain expression.
    No escape character of regular expression is required.
    Domain is checked exact matching or domain boundary suffix matching,
    as config settings.

It is a practical use that making use of both the milter-blockdomain and
the milter-greylist.
The milter-greylist works for unsolved IP address, malformed HELO/EHLO,
GeoIP, widely applied rule by regular expression, and/or S25R.
It is good that the milter-blockdomain handle undesirable domain which is
allowed by auto-whitelist of the milter-greylist.
The milter-regex is also useful.

Equivalent block list expression

    milter-blockdomain

        exapmle.com

    milter-greylist

        racl blacklist domain example.com
        racl blacklist helo /^(.+\.)?exapmle\.com$/
        racl blacklist from /@(.+\.)?exapmle\.com>$/

    milter-regex

        connect /^(.+\.)?exapmle\.com$/ie // and envrcpt /^<.*>$/
        helo    /^(.+\.)?exapmle\.com$/ie    and envrcpt /^<.*>$/
        envfrom /@(.+\.)?exapmle\.com>$/ie   and envrcpt /^<.*>$/

+-------------------+
|   INSTALL FILES   |
+-------------------+

/etc/mail/milter-blockdomain.conf.sample
/usr/local/sbin/milter-blockdomain

+------------------+
|   INSTALLATION   |
+------------------+

1. Build milter-blockdomain

    ./configure --sysconfdir=/etc/mail
    make
    su
    make install

2. Edit your conf file

    cp /etc/mail/milter-blockdomain.conf.sample /etc/mail/milter-blockdomain.conf
    vi /etc/mail/milter-blockdomain.conf

3. Create directory for domain lists

    mkdir /var/lib/milter-blockdomain

4. Create directory for socket file

    mkkdir /var/spool/milter
    chown smmsp:smmsp /var/spool/milter
    chmod o-rwx /var/spool/milter

5. Add "INPUT_MAIL_FILTER" to sendmail.mc

    INPUT_MAIL_FILTER(`milter-blockdomain',`S=local:/var/spool/milter/milter-blockdomain.sock, F=, T=S:1m;R:1m')dnl

6. Generate sendmail.cf from sendmail.mc

    Save current sendmail.cf as a backup.
    cp -p /etc/mail/sendmail.cf /etc/mail/sendmail.cf-yyyymmdd-hhmm

    Change working directory
    cd /usr/share/sendmail/cf/cf

    Generate sendmail.cf from sendmail.mc
    m4 /etc/mail/sendmail.mc | cat -s > /etc/mail/sendmail.cf

+-----------------+
|   MAINTENANCE   |
+-----------------+

1. Edit

    configure.ac
        AC_INIT and AC_DEFINE

2. Autotool

    autoreconf --install --force



ChangeLog

---------------------
2022-04-24  Ver. 1.05
[Enhancement] Command line option -t
              Minimum interval of checking domain list file update.
[Enhancement] Statistics information to syslog daily.
              Command line option -s for hourly statistics info.
[Add] README.md

---------------------
2018-01-20  Ver. 1.04
[Change] Return previous TEMPFAIL or REJECT check result in the SMTP session preferentially.
         Add debug log when the mlfi_xxxx's return.
         milter-blockdomain.c
[Change] Change leaf node anchors have every two characters of the
         beginning of the domain node name, from they have every one
         character of the beginning of the domain node name.
         milter-blockdomain.h  blkdom_check.c  blkdom_config.c
[Change] Add following statistics information,
         Node/Leaf count, Node/Leaf average chain length.
         milter-blockdomain.h - domainListType
         blkdom_config.c - statDomainList, LoadDomainList

---------------------
2017-12-23  Ver. 1.03
[Bug fix] Avoid multiple reloading domain lists in race conditions.
          blkdom_config.c - initConf, ReloadAllDomainList

---------------------
2016-11-03  Ver. 1.02
[Bug fix] Lost the first entry of unsorted domain list.
          Patched: blkdom_config.c - addDomain

---------------------
2016-07-02  Ver. 1.01
[Enhancement] Support GNU build procedure, ./configure, make, and make intall.
              Add config.h generated by the Autoconf/Automake.
[Enhancement] Conf file new keyword 'syslogDebug'. Optional writing syslog priority=debug.
[Enhancement] Conf file new keyword 'checkMacro_'. Check value of the sendmail macro '_'.
[Change] Return SMFIS_CONTINUE when smfi_getpriv returns NULL for the private data pointer.

---------------------
2016-06-05  Ver. 1.00


src/blkdom_check.c
/******************************************************************************/
/*                                                                            */
/*  Copyright(C) 2016, Takao Abe.  All rights reserved.                       */
/*                                                                            */
/*  LICENSE: GPLv3 or any later version                                       */
/*                                                                            */
/******************************************************************************/
/*                                                                            */
/*  ChangeLog                                                                 */
/*                                                                            */
/*  2016/05/21  1.00                                                          */
/*      First release                                                         */
/*                                                                            */
/*  2016/07/02  1.01                                                          */
/*      [Enhancement] Support GNU build procedure.                            */
/*                    Add config.h generated by the Autoconf/Automake.        */
/*                                                                            */
/*  2018/01/14  1.04                                                          */
/*      [Change] Change leaf node anchors have every two characters of the    */
/*               beginning of the domain node name, from they have every one  */
/*               character of the beginning of the domain node name.          */
/*               milter-blockdomain.h  blkdom_check.c  blkdom_config.c        */
/*                                                                            */
/******************************************************************************/

#include "config.h"	/* GNU Autotools */

#include <stdio.h>
#include <stdlib.h>
#include <string.h>
#include <sys/stat.h>

#include "milter-blockdomain.h"

/* External data */

extern	confType xConfData ;
extern	domainListType	xDomainList[ MAX_DOMAIN_LIST ] ;

/**************************************************************************************************/

static bool checkDomain( bool bSubdomain, domainNodeType *pDomainRoot, char *pDomainNodeName[], const int iDomainLevel )
{

	domainLeafType	*pLeafEntry ;
	domainNodeType	*pNodeEntry, *pParentNode ;
	int 	i, iFirstLetterIndex, iSecondLetterIndex ;

	pParentNode = pDomainRoot ;

	for ( i = 0 ; i < iDomainLevel ; i ++ ) {

		iFirstLetterIndex = GET_INDEX( *(pDomainNodeName[i]) ) ;
		iSecondLetterIndex = GET_INDEX( *(pDomainNodeName[i]+1) ) ;

		if ( bSubdomain || i == iDomainLevel - 1 ) {
			pLeafEntry = pParentNode->pLeafList[iFirstLetterIndex][iSecondLetterIndex] ;
			while ( pLeafEntry != NULL ) {
				if ( strcmp( pDomainNodeName[i], pLeafEntry->pName ) == 0 ) {
					/* Found */
					if ( xConfData.bDebugCheckFunction ) {
						logDebug( "%s: Level=%d (Leaf) %s : Match", __func__, i + 1, pLeafEntry->pName ) ;
					}
					return TRUE ;
				} else if ( strcmp( pDomainNodeName[i], pLeafEntry->pName ) < 0 ) {
					/* Not found */
					if ( i == iDomainLevel - 1 ) return FALSE ;
					break ;
				}
				pLeafEntry = pLeafEntry->pNext ;
			}
			if ( i == iDomainLevel - 1 ) return FALSE ;
		}

		pNodeEntry = pParentNode->pNodeList[iFirstLetterIndex] ;
		while ( pNodeEntry != NULL ) {
			if ( strcmp( pDomainNodeName[i], pNodeEntry->pName ) == 0 ) {
				/* Found */
				if ( xConfData.bDebugCheckFunction ) {
					logDebug( "%s: Level=%d (Node) %s : Match", __func__, i + 1, pNodeEntry->pName ) ;
				}
				pParentNode = pNodeEntry ;
				break ;
			} else if ( strcmp( pDomainNodeName[i], pNodeEntry->pName ) < 0 ) {
				/* Not found */
				pNodeEntry = NULL ;
				break ;
			}
			pNodeEntry = pNodeEntry->pNext ;
		}
		if ( pNodeEntry == NULL ) return FALSE ;

	}

	return FALSE ;

}

/**************************************************************************************************/

int checkFQDN( const char *sFQDN )
{

	int 	i ;
	char	sDomainName[ MAX_DOMAIN_NAME_LENGTH + 1 ] ;
	int 	iLength ;
	char	*pDomainNodeName[ MAX_DOMAIN_LEVEL ] ;
	int 	iDomainLevel ;

	if ( sFQDN == NULL ) return RC_NOTLISTED ;
	if ( *sFQDN == 0   ) return RC_NOTLISTED ;

	strncpy( sDomainName, sFQDN, MAX_DOMAIN_NAME_LENGTH ) ;
	sDomainName[MAX_DOMAIN_NAME_LENGTH] = 0 ;
	strLower( sDomainName ) ;
	iLength = strlen( sDomainName ) ;

	if ( sDomainName[0] == '[' && sDomainName[iLength-1] == ']' ) return RC_NOTLISTED ;

	/* Parse domain level */
	iDomainLevel = 0 ;
	for  ( i = iLength - 1 ; i >= 0 ; i -- ) {
		if ( sDomainName[i] == '.' || i == 0 ) {

			if ( iDomainLevel == MAX_DOMAIN_LEVEL ) {
				return RC_NOTLISTED ;
			}

			if ( sDomainName[i] == '.' ) pDomainNodeName[iDomainLevel] = sDomainName + i + 1 ;
			else                         pDomainNodeName[iDomainLevel] = sDomainName + i ;

			if ( strlen( pDomainNodeName[iDomainLevel] ) == 0 ) {
				return RC_NOTLISTED ;
			}

			iDomainLevel ++ ;
			if ( sDomainName[i] == '.' ) sDomainName[i] = 0 ;

		}
	}

	for ( i = 0 ; i < MAX_DOMAIN_LIST ; i ++ ) {
		if ( xDomainList[i].sFileName == NULL ) break ;

		if ( xConfData.bDebugCheckFunction ) logDebug( "%s: %s %s", __func__, sFQDN, xDomainList[i].sFileName ) ;

		if ( checkDomain( xDomainList[i].bSubdomain, &(xDomainList[i].oDomainRoot), pDomainNodeName, iDomainLevel ) ) {
			switch ( xDomainList[i].iListType ) {
			case WHITELIST : return RC_WHITELIST ;
			case BLACKLIST : return RC_BLACKLIST ;
			case GREYLIST  : return RC_GREYLIST ;
			}
		}

	}

	return RC_NOTLISTED ;

}

/**************************************************************************************************/

int checkConnect( const char *sConnect )
{

	int 	rc ;

	rc = checkFQDN( sConnect ) ;
	if ( xConfData.bDebugCheckFunction ) logDebug( "%s: return=%d sConnect=%s", __func__, rc, sConnect ) ;
	return checkFQDN( sConnect ) ;

}

/**************************************************************************************************/

int checkHelo( const char *sHelo )
{

	int 	rc ;

	rc = checkFQDN( sHelo ) ;
	if ( xConfData.bDebugCheckFunction ) logDebug( "%s: return=%d sHelo=%s", __func__, rc, sHelo ) ;
	return rc ;

}

/**************************************************************************************************/

int checkEnvFrom( const char *sEnvFrom )
{

	const char	*pCharLT, *pCharGT, *pCharAtmark ;
	char	sFQDN[ MAX_DOMAIN_NAME_LENGTH + 1 ] ;
	int 	iLength ;
	int 	rc ;

	if ( sEnvFrom == NULL ) return RC_NOTLISTED ;
	if ( *sEnvFrom == 0   ) return RC_NOTLISTED ;

	pCharLT = strchr( sEnvFrom, '<' ) ;
	if ( pCharLT == NULL ) return RC_NOTLISTED ;

	pCharGT = strchr( pCharLT, '>' ) ;
	if ( pCharGT == NULL ) return RC_NOTLISTED ;

	if ( strcmp( pCharLT, "<>" ) == 0 ) return RC_NOTLISTED ;

	pCharAtmark = strchr( pCharLT, '@' ) ;
	if ( pCharAtmark == NULL ) pCharAtmark = pCharLT ;

	strncpy( sFQDN, pCharAtmark + 1, MAX_DOMAIN_NAME_LENGTH ) ;
	sFQDN[MAX_DOMAIN_NAME_LENGTH] = 0 ;
	iLength = strlen( sFQDN ) ;
	if ( sFQDN[iLength-1] == '>' ) sFQDN[iLength-1] = 0 ;

	rc = checkFQDN( sFQDN ) ;
	if ( xConfData.bDebugCheckFunction ) logDebug( "%s: return=%d sFQDN=%s sEnvFrom=%s", __func__, rc, sFQDN, sEnvFrom ) ;
	return rc ;

}

/**************************************************************************************************/

int checkEnvRcpt( const char *sEnvRcpt )
{

	const char	*pCharLT, *pCharGT, *pCharAtmark ;
	char	sFQDN[ MAX_DOMAIN_NAME_LENGTH + 1 ] ;
	int 	iLength ;
	int 	rc ;

	if ( sEnvRcpt == NULL ) return RC_NOTLISTED ;
	if ( *sEnvRcpt == 0   ) return RC_NOTLISTED ;

	pCharLT = strchr( sEnvRcpt, '<' ) ;
	if ( pCharLT == NULL ) return RC_NOTLISTED ;

	pCharGT = strchr( pCharLT, '>' ) ;
	if ( pCharGT == NULL ) return RC_NOTLISTED ;

	if ( strcmp( pCharLT, "<>" ) == 0 ) return RC_NOTLISTED ;

	pCharAtmark = strchr( pCharLT, '@' ) ;
	if ( pCharAtmark == NULL ) pCharAtmark = pCharLT ;

	strncpy( sFQDN, pCharAtmark + 1, MAX_DOMAIN_NAME_LENGTH ) ;
	sFQDN[MAX_DOMAIN_NAME_LENGTH] = 0 ;
	iLength = strlen( sFQDN ) ;
	if ( sFQDN[iLength-1] == '>' ) sFQDN[iLength-1] = 0 ;

	rc = checkFQDN( sFQDN ) ;
	if ( xConfData.bDebugCheckFunction ) logDebug( "%s: return=%d sFQDN=%s sEnvRcpt=%s", __func__, rc, sFQDN, sEnvRcpt ) ;
	return rc ;

}

/**************************************************************************************************/


src/milter-blockdomain.h
/******************************************************************************/
/*                                                                            */
/*    Copyright(C) 2016, Takao Abe.  All rights reserved.                     */
/*                                                                            */
/*    LICENSE: GPLv3 or any later version                                     */
/*                                                                            */
/******************************************************************************/
/*                                                                            */
/*  ChangeLog                                                                 */
/*                                                                            */
/*  2016/05/21  1.00                                                          */
/*      First release                                                         */
/*                                                                            */
/*  2016/07/02  1.01                                                          */
/*      [Enhancement] Conf file new keyword 'syslogDebug'.                    */
/*                    Optional writing syslog priority=debug.                 */
/*      [Enhancement] Conf file new keyword 'checkMacro_'.                    */
/*                    Check value of the sendmail macro '_'.                  */
/*                                                                            */
/*  2018/01/14  1.04                                                          */
/*      [Change] Change leaf node anchors have every two characters of the    */
/*               beginning of the domain node name, from they have every one  */
/*               character of the beginning of the domain node name.          */
/*               milter-blockdomain.h  blkdom_check.c  blkdom_config.c        */
/*      [Change] Add following statistics information,                        */
/*               Node/Leaf count, Node/Leaf average chain length.             */
/*               milter-blockdomain.h - domainListType                        */
/*               blkdom_config.c - statDomainList, LoadDomainList             */
/*                                                                            */
/******************************************************************************/

#ifndef bool
#ifdef	_Bool
#define bool	_Bool
#else
#define bool	char
#endif
#endif

#ifndef TRUE
#define TRUE	(1==1)
#endif

#ifndef FALSE
#define FALSE	(!(TRUE))
#endif

#define RC_SUCCESS 0
#define RC_ERROR   1

#define MAX_CONF_LINE_LENGTH	100

#define MAX_DOMAIN_LIST         8
#define MAX_DOMAIN_NAME_LENGTH  255
#define MAX_DOMAIN_LEVEL        128

typedef struct domainLeafStruct
{
	char	*pName ;
	short	iLength ;
	struct  domainLeafStruct	*pNext ;
} domainLeafType ;

#define MAX_DOMAIN_INDEX	40
#define GET_INDEX(ch)	( ch - ( ch > 'z' ? (ch-38) : ( ch >= 'a' ? 85 : ( ch > 'Z' ? (ch-38) : ( ch >= 'A' ? 53 : ( ch > '9' ? (ch-11) : ( ch >= '0' ? 47 : ch ) ) ) ) ) ) )

typedef struct domainNodeStruct
{
	char	*pName ;
	short	iLength ;
	short	iLevel ;
	struct  domainLeafStruct	*pLeafList[MAX_DOMAIN_INDEX][MAX_DOMAIN_INDEX] ;
	struct  domainNodeStruct	*pNodeList[MAX_DOMAIN_INDEX] ;
	struct  domainNodeStruct	*pNext ;
} domainNodeType ;

typedef struct
{
	char	*sFileName ;
	struct  stat	xFileStat ;
	enum { WHITELIST, GREYLIST, BLACKLIST }	iListType ;
	bool	bSubdomain ;
	domainNodeType	oDomainRoot ;
	int 	iStatTotal, iStatTLD, iStatDepth, iStatMaxLeafChain, iStatMaxNodeChain ;
	int 	iStatNumberOfLeafAnchors, iStatTotalLeafChain, iStatAverageLeafChain ;
	int 	iStatNumberOfNodeAnchors, iStatTotalNodeChain, iStatAverageNodeChain ;
} domainListType ;

typedef struct
{
	char	*sUser ;
	char	*sSocket ;
	char	*sPidFile ;
	bool	bSubdomain ;
	bool	bCheckConnect ;
	bool	bCheckMacro_ ;
	bool	bCheckHelo ;
	bool	bCheckEnvFrom ;
	bool	bCheckEnvRcpt ;
	short	iCheckOnAndAfter ;	/* CHECK_ON_AND_AFTER_XXXX */
	char	*sMessageTempfail ;
	char	*sMessageReject ;
	char	*pSyslogFacility ;
	bool	bSyslogDebug ;
	/* Test mode */
	bool	bTestMode ;
	/* Debug log keyword */
	bool	bDebugConfigFunction ;
	bool	bDebugCheckFunction ;
} confType ;

#define CHECKPOINT_CONNECT	0
#define CHECKPOINT_HELO   	1
#define CHECKPOINT_ENVFROM	2
#define CHECKPOINT_ENVRCPT	3
#define CHECKPOINT_HEADER 	4
#define CHECKPOINT_BODY   	5

/* Function prototype ( blkdom_util.c ) */

void setLogLevel  ( const int ) ;
void setSyslogDebug( const bool ) ;
void openSyslog( const char *, const char * ) ;
void logDebug  ( const char *, ... ) ;
void logInfo   ( const char *, ... ) ;
void logWarning( const char *, ... ) ;
void logError  ( const char *, ... ) ;
void logAlert  ( const char *, ... ) ;

void strLower  ( char * ) ;

/* Function prototype ( blkdom_config.c ) */

void initConf() ;
void freeConf() ;
int  LoadConf( const char * ) ;
int  LoadAllDomainList() ;
int  ReloadAllDomainList() ;
void freeAllDomainList() ;

/* Function prototype ( blkdom_check.c ) */

#define RC_NOTLISTED	0
#define RC_WHITELIST	1
#define RC_BLACKLIST	2
#define RC_GREYLIST 	3

int checkConnect( const char * ) ;
int checkHelo   ( const char * ) ;
int checkEnvFrom( const char * ) ;
int checkEnvRcpt( const char * ) ;

/* Log message */

#define ALERT_MESSAGE_STRDUP_NULL       "$s: Memory allocation error ( strdup )"
#define ALERT_MESSAGE_MALLOC_NULL       "%s: Memory allocation error ( malloc )"
#define ALERT_MESSAGE_SOCKET_ERROR      "%s: smfi_setconn error %s"
#define ALERT_MESSAGE_MILTER_FUNCTION   "%s: %s error"
#define ALERT_MESSAGE_DAEMON            "%s: Could not be daemon"

#define ERROR_MESSAGE_FILE_OPEN         "Open error %s"
#define ERROR_MESSAGE_FILE_STAT         "Stat error %s"
#define ERROR_MESSAGE_FILE_UNLINK       "Unlink error %s"
#define ERROR_MESSAGE_INET_NTOP         "inet_ntop error"
#define ERROR_MESSAGE_UNKNOWN_AF        "Unknown address family"
#define ERROR_MESSAGE_GETPWNAM          "User not found '%s'"
#define ERROR_MESSAGE_SETGID            "setgid error: %s gid=%d"
#define ERROR_MESSAGE_SETUID            "setuid error: %s uid=%d"
#define ERROR_MESSAGE_CHOWN             "chown error: %s uid=%d gid=%d"

#define ERROR_MESSAGE_INVALID_CONFIG_KEYWORD    	"Invalid config file: Line=%d [%s] : Invalid keyword"
#define ERROR_MESSAGE_INVALID_CONFIG_MISSING_1  	"Invalid config file: Line=%d [%s] [] : Missing value"
#define ERROR_MESSAGE_INVALID_CONFIG_MISSING_2  	"Invalid config file: Line=%d [%s] [%s] [] : Missing value"
#define ERROR_MESSAGE_INVALID_CONFIG_VALUE_1    	"Invalid config file: Line=%d [%s] [%s] : Invalid value"
#define ERROR_MESSAGE_INVALID_CONFIG_VALUE_2    	"Invalid config file: Line=%d [%s] [%s] [%s] : Invalid value"
#define ERROR_MESSAGE_INVALID_CONFIG_MAX_LIST   	"Invalid config file: Line=%d [%s] : Exceed max domain list"
#define ERROR_MESSAGE_INVALID_CONFIG_ERROR_COUNT	"Invalid config file: Error count = %d"

#define ERROR_MESSAGE_INVALID_DOMAINLIST_NODE_LENGTH	"Invalid domain list: Line=%d [%s] : Continuous dot"
#define ERROR_MESSAGE_INVALID_DOMAINLIST_NODE_DEPTH 	"Invalid domain list: Line=%d [%s] : Too deep domain level"

#define	WARNING_MESSAGE_INVALID_DOMAINLIST_ROOTDOMAIN	"Invalid domain list: Line=%d [%s] : Root domain"
#define	WARNING_MESSAGE_INVALID_DOMAINLIST_DUPLICATE	"Invalid domain list: Line=%d [%s] : Duplicate"

#define INFO_MESSAGE_PROGRAM_VERSION	"%s  %s  ( %s ), built on %s"
#define INFO_MESSAGE_CONFIG	"Config file %s"
#define INFO_MESSAGE_STOP	"Stop"



src/blkdom_util.c
/******************************************************************************/
/*                                                                            */
/*  Copyright(C) 2016, Takao Abe.  All rights reserved.                       */
/*                                                                            */
/*  LICENSE: GPLv3 or any later version                                       */
/*                                                                            */
/******************************************************************************/
/*                                                                            */
/*  ChangeLog                                                                 */
/*                                                                            */
/*  2016/05/21  1.00                                                          */
/*      First release                                                         */
/*                                                                            */
/*  2016/07/02  1.01                                                          */
/*      [Enhancement] Support GNU build procedure.                            */
/*                    Add config.h generated by the Autoconf/Automake.        */
/*      [Enhancement] Conf file new keyword 'syslogDebug'.                    */
/*                    Optional writing syslog priority=debug.                 */
/*                                                                            */
/******************************************************************************/

#include "config.h"	/* GNU Autotools */

#include <stdio.h>
#include <stdlib.h>
#include <string.h>
#include <stdarg.h>
#include <syslog.h>
#include <ctype.h>
#include <sys/stat.h>

#include "milter-blockdomain.h"

/* Local data */

static	int 	iDebugLevel = 0 ;
static	bool	bSyslogDebug = FALSE ;

/**************************************************************************************************/

void setLogLevel( const int iLogLevel )
{
	iDebugLevel = iLogLevel ;
}

/**************************************************************************************************/

void setSyslogDebug( const bool bFlag )
{
	bSyslogDebug = bFlag ;
}

/**************************************************************************************************/

void openSyslog( const char *pProcessName, const char *pFacilityName )
{

	int 	iFacility = LOG_MAIL ;

	if      ( pFacilityName == NULL                  ) iFacility = LOG_MAIL   ;
	else if ( strcmp( pFacilityName, "daemon" ) == 0 ) iFacility = LOG_DAEMON ;
	else if ( strcmp( pFacilityName, "mail"   ) == 0 ) iFacility = LOG_MAIL   ;
	else if ( strcmp( pFacilityName, "user"   ) == 0 ) iFacility = LOG_USER   ;
	else if ( strcmp( pFacilityName, "local0" ) == 0 ) iFacility = LOG_LOCAL0 ;
	else if ( strcmp( pFacilityName, "local1" ) == 0 ) iFacility = LOG_LOCAL1 ;
	else if ( strcmp( pFacilityName, "local2" ) == 0 ) iFacility = LOG_LOCAL2 ;
	else if ( strcmp( pFacilityName, "local3" ) == 0 ) iFacility = LOG_LOCAL3 ;
	else if ( strcmp( pFacilityName, "local4" ) == 0 ) iFacility = LOG_LOCAL4 ;
	else if ( strcmp( pFacilityName, "local5" ) == 0 ) iFacility = LOG_LOCAL5 ;
	else if ( strcmp( pFacilityName, "local6" ) == 0 ) iFacility = LOG_LOCAL6 ;
	else if ( strcmp( pFacilityName, "local7" ) == 0 ) iFacility = LOG_LOCAL7 ;

	openlog( pProcessName, LOG_PID, iFacility ) ;

}

/**************************************************************************************************/

void logDebug( const char *format, ... )
{

	va_list	arg ;

	if ( bSyslogDebug ) {
		va_start( arg, format ) ;
		vsyslog( LOG_DEBUG, format, arg ) ;
		va_end( arg ) ;
	}

	if ( iDebugLevel > 0 ) {
		va_start( arg, format ) ;
		vprintf( format, arg ) ;
		va_end( arg ) ;
		printf( "\n" ) ;
	}

}

/**************************************************************************************************/

void logInfo( const char *format, ... )
{

	va_list	arg ;

	va_start( arg, format ) ;
	vsyslog( LOG_INFO, format, arg ) ;
	va_end( arg ) ;

	if ( iDebugLevel > 0 ) {
		va_start( arg, format ) ;
		vprintf( format, arg ) ;
		va_end( arg ) ;
		printf( "\n" ) ;
	}

}

/**************************************************************************************************/

void logWarning( const char *format, ... )
{

	va_list	arg ;

	va_start( arg, format ) ;
	vsyslog( LOG_WARNING, format, arg ) ;
	va_end( arg ) ;

	if ( iDebugLevel > 0 ) {
		va_start( arg, format ) ;
		vprintf( format, arg ) ;
		va_end( arg ) ;
		printf( "\n" ) ;
	}

}

/**************************************************************************************************/

void logError( const char *format, ... )
{

	va_list	arg ;

	va_start( arg, format ) ;
	vsyslog( LOG_ERR, format, arg ) ;
	va_end( arg ) ;

	if ( iDebugLevel > 0 ) {
		va_start( arg, format ) ;
		vprintf( format, arg ) ;
		va_end( arg ) ;
		printf( "\n" ) ;
	}

}

/**************************************************************************************************/

void logAlert( const char *format, ... )
{

	va_list	arg ;

	va_start( arg, format ) ;
	vsyslog( LOG_ALERT, format, arg ) ;
	va_end( arg ) ;

	if ( iDebugLevel > 0 ) {
		va_start( arg, format ) ;
		vprintf( format, arg ) ;
		va_end( arg ) ;
		printf( "\n" ) ;
	}

}

/**************************************************************************************************/

void strLower( char *s )
{

	for ( ; *s != 0 ; s ++ ) *s = tolower( *s ) ;

}

/**************************************************************************************************/
